xrtl/gfx/util/memory_command_decoder.cc

```diff
@@ -0,0 +1,485 @@
+        break;
+      }
+      case CommandType::kCopyImage: {
+        auto command = reader->ReadCommand<CopyImageCommand>(command_header);
+        auto regions = reader->ReadArray<CopyImageRegion>(command.region_count);
+        any_transfer_encoder->CopyImage(ref_ptr<Image>(command.source_image),
+                                        command.source_image_layout,
+                                        ref_ptr<Image>(command.target_image),
+                                        command.target_image_layout, regions);
+        break;
+      }
+      case CommandType::kCopyBufferToImage: {
+        auto command =
+            reader->ReadCommand<CopyBufferToImageCommand>(command_header);
+        auto regions =
+            reader->ReadArray<CopyBufferImageRegion>(command.region_count);
+        any_transfer_encoder->CopyBufferToImage(
+            ref_ptr<Buffer>(command.source_buffer),
+            ref_ptr<Image>(command.target_image), command.target_image_layout,
+            regions);
+        break;
+      }
+      case CommandType::kCopyImageToBuffer: {
+        auto command =
+            reader->ReadCommand<CopyImageToBufferCommand>(command_header);
+        auto regions =
+            reader->ReadArray<CopyBufferImageRegion>(command.region_count);
+        any_transfer_encoder->CopyImageToBuffer(
+            ref_ptr<Image>(command.source_image), command.source_image_layout,
+            ref_ptr<Buffer>(command.target_buffer), regions);
+        break;
+      }
+      case CommandType::kBlitImage: {
+        auto command = reader->ReadCommand<BlitImageCommand>(command_header);
+        auto regions = reader->ReadArray<BlitImageRegion>(command.region_count);
+        render_encoder->BlitImage(
+            ref_ptr<Image>(command.source_image), command.source_image_layout,
+            ref_ptr<Image>(command.target_image), command.target_image_layout,
+            command.scaling_filter, regions);
+        break;
+      }
+      case CommandType::kResolveImage: {
+        auto command = reader->ReadCommand<ResolveImageCommand>(command_header);
+        auto regions = reader->ReadArray<CopyImageRegion>(command.region_count);
+        render_encoder->ResolveImage(ref_ptr<Image>(command.source_image),
+                                     command.source_image_layout,
+                                     ref_ptr<Image>(command.target_image),
+                                     command.target_image_layout, regions);
+        break;
+      }
+      case CommandType::kGenerateMipmaps: {
+        auto command =
+            reader->ReadCommand<GenerateMipmapsCommand>(command_header);
+        render_encoder->GenerateMipmaps(ref_ptr<Image>(command.image));
+        break;
+      }
+
+      case CommandType::kClearColorImage: {
+        auto command =
+            reader->ReadCommand<ClearColorImageCommand>(command_header);
+        auto ranges = reader->ReadArray<Image::LayerRange>(command.range_count);
+        if (compute_encoder) {
```

Comment:
same down here (assumes that if compute_encoder isn't set then render_encoder is)